microovn/frr/bgp/service.go

```diff
@@ -22,9 +39,35 @@ func EnableService() error {
 		logging.Errorf("Failed to start %s service: %s", FrrBgpService, err)
 		return errors.New("failed to start BGP service")
 	}
-	return nil
+
+	if extraConfig == nil {
+		return nil
+	}
+
+	extConnections, err := extraConfig.ParseExternalConnection()
+	if err != nil {
+		logging.Errorf("Failed to parse external connections: %v", err)
+	}
+
+	err = createExternalBridges(ctx, s, extConnections)
+	if err != nil {
+		return err
+	}
+
+	err = createExternalNetworks(ctx, s, extConnections)
+	if err != nil {
+		return err
+	}
+
+	err = createVrf(ctx, s, extConnections, extraConfig.Vrf)
+	if err != nil {
+		return err
+	}
+
+	return redirectBgp(ctx, s, extConnections, extraConfig.Vrf)
 }
```

Comment:
I don't see us creating the initial BGP daemon config here, which I would expect.

I know this part is less clearly defined, but approaching it will give us much information about what we need to consider in the final implementation.

Please provide a oneshot type call populating the bare minimum similar to what the functional test does:
```
router bgp 1 vrf ovnvrf10
 neighbor eth1-bgp interface remote-as internal
 neighbor eth2-bgp interface remote-as internal
exit
```

Reply:
I would disagree on this point. If we commit to doing this, we will need to think about:
* How to properly interact with FRR from the MicroOVN codebase
* Do we want to save the running config when we add this?
  * If yes, we need to purge the config on `microovn bgp disable` (or somehow selectively only remove the section we added)
  * If not, do we want to re-create this config on service restart/reboot
* Additional config option in for local ASN (possibly ASNs also for `remote-as`)
* Does user want bgp `unnumbered` or regular explicit peers?

I think that either we are going to make the `microovn enable` config options way too complicated, or we are going to enforce some config options on the user that they may not want.

Overall I think that the upside of us getting involved with the BGP daemon configuration is very little, as any user that wants this feature should be capable of configuring BGP daemon on their own, and the downside is significant in terms of complexity on the code.

Reply:
After our follow-up offline conversation, I added `asn` config option to `microovn enable bgp` command. When supplied, MicroOVN will set up iBGP unnumbered FRR config. As for the questions I laid out in my previous comment:

> How to properly interact with FRR from the MicroOVN codebase

This actually turned out to be quite simple via executing `vtysh` with multiple `-c` options, effectively achieving "script-like" execution inside the FRR shell.

> Do we want to save the running config when we add this?

yes, it's much easier than trying to re-inject correct config on service restart

> If yes, we need to purge the config on microovn bgp disable (or somehow selectively only remove the section we added)

On disable, current "startup" config is backed up and then replaced with default frr config.

> Additional config option in for local ASN (possibly ASNs also for remote-as)

`asn` option was added and all BGP interfaces are lumped into this ASN, there's no option to define per-interface ASN now. As for `remote-as`, we always use `internal`

> Does user want bgp unnumbered or regular explicit peers?

User can omit the `asn` config option, in which case MicroOVN won't be configuring anything in FRR. This allows users to configure FRR manually if the auto-generated config from MicroOVN wouldn't suite their needs.